Modified README to include Model Registry component; augmented the Mo… (#2493)
mlflow/mlflow: docs/source/model-registry.rst

Concepts
========

> As the fourth standalone component of MLflow, the Model Registry introduces few concepts that describe and facilitate the full lifecycle of an MLflow model.

Reviewer: ```suggestion
As the fourth standalone component of MLflow, the Model Registry introduces a few concepts that describe and facilitate the full lifecycle of an MLflow model.
```

Reviewer: Is there a reason why we are highlighting that this is the "fourth" component?

Author: More from having had three components for a while and having MLflow touted as component-based and modularized, thought we would highlight it. The next or 5th pillar might be Model Serving. I

Author: Thinking more about it, "standalone" may be misleading or ambiguous as much as I wish to convey its modularity, and "fourth" I'm not wedded to it. We can drop both, and just rephrase as
"As a new component of MLflow," or just drop the introductory phrase. Let you decide.

Reviewer: I vote for dropping the phrase.

Model
    An MLflow Model is created from an experiment or run that is logged with one of the model flavor’s ``mlflow.model_flavor.log_model`` methods. Once logged, this model can then be registered with the Model Registry.

Registered Model

>     An MLflow Model can be registered with the  Model Registry. It has a unique name, contains versions, associated transitional stages, model lineage, and other metadata.

Reviewer: ```suggestion
    An MLflow Model can be registered with the  Model Registry. A registered model has a unique name, contains versions, associated transitional stages, model lineage, and other metadata.
```

Reviewer: Be careful about replacing nouns with pronouns. In your previous version, it's not clear if "It" refers to the Model Registry or the Model.

Model Version
    Each registered model can have one or many versions. When a new model is added to the Model Registry, it is added as version 1. Each new model registered to the same model name increments the version number.

Model Stage
    Each distinct model version can be assigned one stage at any given time. MLflow provides predefined stages for common use-cases such as *Staging*, *Production* or *Archived*. You can transition a model version from one stage to another stage.

Annotations and Descriptions

>     You can annotate the top-level model and each version individually using markdown, including description and any relevant information useful for the team such as algorithm descriptions, dataset employed or methodology.

Reviewer: ```suggestion
    You can annotate the top-level model and each version individually using Markdown, including description and any relevant information useful for the team such as algorithm descriptions, dataset employed or methodology.
```

Model Registry Workflows
========================

> Before a model can be added to the Model Registry you must log it using the ``log_model`` methods

Reviewer: ```suggestion
Before you can add a model to the Model Registry you must log it using the ``log_model`` methods
```

Concepts
========

The Model Registry introduces a few concepts that describe and facilitate the full lifecycle of an MLflow Model.

Model

>     An MLflow Model is created from an experiment or run that is logged with one of the model flavor’s ``mlflow.model_flavor.log_model`` methods. Once logged, this model can then be registered with the Model Registry.

Reviewer 2: Maybe `mlflow.<model_flavor>.log_model` to help clarify that it's not a real snippet that can run?

Author: Yes, indeed, that bit is true. Will change it.

In particular, you can register a model during an MLflow experiment run or after all your experiment runs.

Adding an MLflow Model to the Model Registry
--------------------------------------------

There are three programmatic ways to add a model to the registry. First, you can use the ``mlflow.model_flavor.log_model(...)`` method. For example, in your code:

.. code-block:: py

    with mlflow.start_run(run_name="YOUR_RUN_NAME") as run:
        params={"n_estimators": 5, "random_state": 42}
        sk_learn_rfr=RandomForestRegressor(params)
        # Log params using the MLflow sklearn flavor APIs
        mlflow.log_params(params)
        log_param("param_1", randint(0, 100))
        log_metric("metric_1", random())
        log_metric("metric_2", random() + 1)
       …
       # Log the sklearn model and register as version 1
       mlflow.sklearn.log_model(sk_model=sk_learn_rfr,
                artifact_path="sklearn-model",
                registered_model_name="sk-learn-random-forest-reg-model")

> This will log the model as well as register it under the specified name as version 1.

Reviewer: ```suggestion
This logs the model as well as registers it under the specified name as version 1.
```

.. code-block:: py

    client=MlflowClient()
    [pprint.pprint(dict(rm), indent=4) for rm in client.list_registered_models()]

    {   'creation_timestamp': 1582671933216,
        'description': None,
        'last_updated_timestamp': 1582671960712,
        'latest_versions': [<ModelVersion: creation_timestamp=1582671933246, current_stage='Production', description='A random forest model containing 100 decision trees trained in scikit-learn', last_updated_timestamp=1582671960712, name='sk-learn-random-forest-reg-model', run_id='ae2cc01346de45f79a44a320aab1797b', source='./mlruns/0/ae2cc01346de45f79a44a320aab1797b/artifacts/sklearn-model', status='READY', status_message=None, user_id=None, version=1>,
                           <ModelVersion: creation_timestamp=1582671960628, current_stage='None', description=None, last_updated_timestamp=1582671960628, name='sk-learn-random-forest-reg-model', run_id='d994f18d09c64c148e62a785052e6723', source='./mlruns/0/d994f18d09c64c148e62a785052e6723/artifacts/sklearn-model', status='READY', status_message=None, user_id=None, version=2>],
        'name': 'sk-learn-random-forest-reg-model'}
    ...
    ...

> With hundreds of models, this call can be cumbersome to peruse. A more appropriate call would be to search a specific model name and list its version

Reviewer: ```suggestion
With hundreds of models, this call can be cumbersome to peruse. A more appropriate call would be to search for a specific model name and list its version
```

In particular, you can register a model during an MLflow experiment run or after all your experiment runs.

Adding an MLflow Model to the Model Registry
--------------------------------------------

There are three programmatic ways to add a model to the registry. First, you can use the ``mlflow.<model_flavor>.log_model(...)`` method. For example, in your code:

.. code-block:: py

    with mlflow.start_run(run_name="YOUR_RUN_NAME") as run:
        params={"n_estimators": 5, "random_state": 42}
        sk_learn_rfr=RandomForestRegressor(params)
        # Log parameters and metrics using the MLflow APIs
        mlflow.log_params(params)
        log_param("param_1", randint(0, 100))
        log_metric("metric_1", random())
        log_metric("metric_2", random() + 1)
       …
       # Log the sklearn model and register as version 1
       mlflow.sklearn.log_model(sk_model=sk_learn_rfr,
                artifact_path="sklearn-model",

>                 registered_model_name="sk-learn-random-forest-reg-model")

Reviewer 3: This `...` are unrecognized characters.

Author: Ah, no wonder it won't highlight the strings... was wondering.
Will remove it.

At any point in a model’s lifecycle development, you can update a model version's description using the MLflow Tracking API.

.. code-block:: py

    client = MlflowClient()
    client.update_model_version(name="sk-learn-random-forest-reg-model",
            version=1,
            description="This model version is a scikit-learn random forest containing 100 decision trees")

As well as adding or updating a description of a specific version of the model, you can rename an existing registered model.

.. code-block:: py

    client=MlflowClient()
    client.rename_registered_model(name="sk-learn-random-forest-reg-model",
            new_name="sk-learn-random-forest-reg-model-100")

Transitioning an MLflow Model’s Stage
-------------------------------------
Over the course of the model’s lifecycle, a model evolves—from development to staging to production.
You can transition a registered model in the registry to one of the stages: **None**, **Staging**, **Production** or **Archived**.

.. code-block:: py

    client = MlflowClient()
    client.transition_model_version_stage(name="sk-learn-random-forest-reg-model",
  	        version=3,
	        stage="production")

> Listing and Searching MLflow Registered Models

Reviewer 3: => Listing and Searching Models

(because `list` an API on registered models and `search` on model versions)

Author: I'm on it. Will commit shortly.

Author: done and addressed it with a new set of commits

API Workflow
============

An alternative way to interact with Model Registry is using the `MLflow model flavor <https://www.mlflow.org/docs/latest/python_api/index.html>`_ or `MLflow Client Tracking API <https://www.mlflow.org/docs/latest/python_api/mlflow.tracking.html>`_ interface.
In particular, you can register a model during an MLflow experiment run or after all your experiment runs.

Adding an MLflow Model to the Model Registry
--------------------------------------------

There are three programmatic ways to add a model to the registry. First, you can use the ``mlflow.model_flavor.log_model(...)`` method. For example, in your code:

.. code-block:: py

    with mlflow.start_run(run_name="YOUR_RUN_NAME") as run:
        params={"n_estimators": 5, "random_state": 42}
        sk_learn_rfr=RandomForestRegressor(params)

>         # Log params using the MLflow sklearn flavor APIs

Reviewer 2: What does this line mean? There is not a flavored param logger.

Author: I meant " Log parameters using the MLflow API"

        mlflow.log_params(params)
        log_param("param_1", randint(0, 100))
        log_metric("metric_1", random())
        log_metric("metric_2", random() + 1)
       …
       # Log the sklearn model and register as version 1
       mlflow.sklearn.log_model(sk_model=sk_learn_rfr,
                artifact_path="sklearn-model",
                registered_model_name="sk-learn-random-forest-reg-model")

This logs the model as well as registers it under the specified name as version 1.

The second way is to explicitly register the `mlflow.register_model(...) <https://www.mlflow.org/docs/latest/python_api/mlflow.html#mlflow.register_model>`_,
after all your experiment runs and when you have ascertained which run within an experiment is most suitable to add to the registry.
For this scheme, you will need the ``run_id`` as part of the ``runs:URI`` argument.

.. code-block:: py

> 

Reviewer 2: Maybe we can name the arguments like above so it's clear what they are referring to

Author: I think "EmailMarketingCampaign" seems more like an Experiment name you want under which will want to create a set of different types of models—scikit-learn flavor, Keras or PyTorch etc. 

I suppose we can all it WeatherPowerForecast as general if you prefer.

>     result=mlflow.register_model(“runs:/d16076a3ec534311817565e6527539c0/artifacts/sk-model”,

Reviewer 2: Are these smart quotes, the bane of all copy-pasting?  Maybe make them regular quotes :)

Author: I made them regular quotes, will try do it again. Yes cut-and-paste hast its hidden gotchas.

As with above ``mlflow.sklearn.log_model(...)``, this method creates version 1 of the specified model and it returns a single `ModelVersion <https://www.mlflow.org/docs/latest/python_api/mlflow.entities.html#mlflow.entities.model_registry.ModelVersion>`_ MLflow object.

And finally, you can use the `MLflow Client Tracking API <https://www.mlflow.org/docs/latest/python_api/mlflow.tracking.html#mlflow.tracking.MlflowClient.create_registered_model>`_ to create a new registered model. If the model name exists, this will throw an ``MLflowException`` since creating a new registered model requires a unique name.

.. code-block:: py

   client = MlflowClient()
   client.create_registered_model("sk-learn-random-forest-reg-model")

While the method above creates an empty registered model with no version associated, the method below creates a new version of the model.

.. code-block:: py

    client = MlflowClient()
    result = client.create_model_version(name="sk-learn-random-forest-reg-model",
                source = "mlruns/0/d16076a3ec534311817565e6527539c0/artifacts/sk-model",
                run_id = "d16076a3ec534311817565e6527539c0")

> In contrast, ``mlflow.register_model()`` and ``mlflow.model_flavor.log_model()`` will create a new version in the registry.

Reviewer 2: , if it does not already exist

Adding or Updating Model Descriptions
-------------------------------------

At any point in a model’s lifecycle development, you can update a model version's description using the MLflow Tracking API.

.. code-block:: py

    client = MlflowClient()
    client.update_model_version(name=”sk-learn-random-forest-reg-model”,
            version=1,

>             description="This model version is a random forest containing 100 decision trees that was trained in scikit-learn.")

Reviewer 2: Since the code snippet does not wrap, maybe make the description a bit shorter.

Author: Wil make is shorter

As well as adding or updating a description of a specific version of the model, you can rename an existing registered model.

.. code-block:: py

    client=MlflowClient()

>     client.rename_registered_model(name=”sk-learn-random-forest-reg-model”,

Reviewer 2: smart quotes

Author: will fix as smart quotes.

            new_name= ”sk-learn-random-forest-reg-model-100”)

Transitioning an MLflow Model’s Stage
-------------------------------------
Over the course of the model’s lifecycle, a model evolves—from development to staging to production.

> You can transition a registered model in the registry to one of the stages: **Staging**, **Production** or **Archived**.

Reviewer 2: None is also possible! You can move a model back to the None stage.

Author: Ah, interesting. Will note that.

Reviewer 3: Model versions added to registry are in `None` stage. The way this is envisioned is versions moving forward from `Staging` --> `Production` or `Archived` directly. While, the APIs do no block moving versions to `None` it makes little sense. 

Let's **not** document transitioning to `None`. Let's keep it as is.

.. code-block:: py

    client = MlflowClient()
    client.transition_model_version_stage(name=”sk-learn-random-forest-reg-model”,
  	        version=3,

> 	        stage=“production”)

Reviewer 2: We should probably include an example of `archive_existing_versions`, right @mparkhe ?

Author: I believe this is only an edge feature. We don't expose that in the OSS.

Reviewer 3: `archive_existing_versions` is not support in SQL backend

    {   'creation_timestamp': 1582671933246,
        'current_stage': 'Production',
        'description': 'A random forest model containing 100 decision trees '
                       'trained in scikit-learn',
        'last_updated_timestamp': 1582671960712,
        'name': 'sk-learn-random-forest-reg-model',
        'run_id': 'ae2cc01346de45f79a44a320aab1797b',
        'source': './mlruns/0/ae2cc01346de45f79a44a320aab1797b/artifacts/sklearn-model',
        'status': 'READY',
        'status_message': None,
        'user_id': None,
        'version': 1}

    {   'creation_timestamp': 1582671960628,
        'current_stage': 'None',
        'description': None,
        'last_updated_timestamp': 1582671960628,
        'name': 'sk-learn-random-forest-reg-model',
        'run_id': 'd994f18d09c64c148e62a785052e6723',
        'source': './mlruns/0/d994f18d09c64c148e62a785052e6723/artifacts/sklearn-model',
        'status': 'READY',
        'status_message': None,
        'user_id': None,
        'version': 2
    }

Deleting Registered Models
--------------------------

> You can either delete a specific version of the registered model or registered model and all its versions.

Reviewer 2: Clause is a bit confusing, maybe repeat "can delete":

`You can either delete a specific version of the registered model or you can delete the registered model and all its versions.`

Author: Will fix it.

Reviewer 3: "Archived" is a stage for a model version is a part of lifecycle stage. Usually people would want to keep archived models around for long. Deleting as you mentioned is permanently deleting the model version from the registry. I am not super psyched about mixing these 2 concepts together. How about this.

```
Archiving Models
----------------

To transition model versions out of **Production** stage, transition those model into **Archived** 
stage.

.. code-block:: py

    client = MlflowClient()
    client.transition_model_version_stage(name="sk-learn-random-forest-reg-model",
  	        version=3,
	          stage="Archived")

Deleting Models
---------------

.. note::
  Deleting registered models and model versions is irrevocable, so use it judiciously.

You can either delete a specific version of a registered model or you can delete a registered model 
and all its versions.

.. code-block:: py

    # Delete models versions 1, 2, and 3
    versions = [1, 2, 3]
    client=MlflowClient()
    for version in versions:
      client.delete_model_version(name="sk-learn-random-forest-reg-model", version=version)

    # Delete the registered model along with all its versions
    client.delete_registered_model(name="sk-learn-random-forest-reg-model")

```

This operation is irrevocable, so use it judiciously.

.. code-block:: py

    client=MlflowClient()
    # Delete a specific version of the model
    client.delete_model_version(name=”sk-learn-random-forest-reg-model”,
            version=1 )
    # Delete the registered model along with all its versions
    client.delete_registered_model(name=”sk-learn-random-forest-reg-model”)

> However, if you are unsure, you may want to archive specific versions of a registered model.

Reviewer 2: Maybe flip these. We actually expect people to archive most often. Maybe the session could be "Archiving and deleting registered models" and we first say how to archive and then say if you want to permanently delete, use this instead.

Author: Makes sense, better to first Archive them, and later they can always delete once they are certain.
